community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.logging;
+
+import java.util.concurrent.Executor;
+
+import org.neo4j.concurrent.AsyncEvents;
+import org.neo4j.function.Consumer;
+import org.neo4j.kernel.impl.util.CappedOperation;
+import org.neo4j.kernel.impl.util.Listener;
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import static org.neo4j.kernel.impl.util.CappedOperation.differentItemClasses;
+import static org.neo4j.kernel.impl.util.CappedOperation.time;
+
+/**
+ * Wraps a {@link LogProvider}, making its {@link Log logs} asynchronous by letting their log statements be
+ * treated as events fed into a {@link AsyncEvents} to be processed by a dedicated logging thread instead
+ * of in the thread making the log statement call.
+ */
+class AsyncLogProvider
+        extends LifecycleAdapter
+        implements LogProvider
+{
+    private final AsyncEvents<AsyncLogEvent> events;
+    private final LogProvider actual;
+    private final Executor scheduler;
+
+    AsyncLogProvider( LogProvider actual, Executor scheduler )
+    {
+        this( actual, scheduler, new Listener<Throwable>()
+        {
+            @Override
+            public void receive( Throwable error )
+            {
+                error.printStackTrace();
+            }
+        } );
+    }
+
+    AsyncLogProvider( LogProvider actual, Executor scheduler, final Listener<Throwable> errorHandler )
+    {
+        this.scheduler = scheduler;
+        this.events = new AsyncEvents<>( new Consumer<AsyncLogEvent>()
```

Comment:
Why do we introduce new structure for AsyncEvents? How performance wise it's comparable with different kinds of j.u.c.* Queues?